From: ThomasKaiser <ThomasKaiser@users.noreply.github.com>
Date: Fri, 1 Jul 2022 11:24:04 +0200
Subject: [PATCH 16/24] Adding Apple core names, distinguish by SoC/SiP name

---
 sys-utils/lscpu-arm.c | 12 +++++++++---
 1 file changed, 9 insertions(+), 3 deletions(-)

diff --git a/sys-utils/lscpu-arm.c b/sys-utils/lscpu-arm.c
index 30977cd..7b9b364 100644
--- a/sys-utils/lscpu-arm.c
+++ b/sys-utils/lscpu-arm.c
@@ -159,9 +159,15 @@ static const struct id_part marvell_part[] = {
 };
 
 static const struct id_part apple_part[] = {
-    { 0x022, "Icestorm" },
-    { 0x023, "Firestorm" },
-    { -1, "unknown" },
+    { 0x020, "Icestorm-T8101" },
+    { 0x021, "Firestorm-T8101" },
+    { 0x022, "Icestorm-T8103" },
+    { 0x023, "Firestorm-T8103" },
+    { 0x030, "Blizzard-T8110" },
+    { 0x031, "Avalanche-T8110" },
+    { 0x032, "Blizzard-T8112" },
+    { 0x033, "Avalanche-T8112" },
+   { -1, "unknown" },
 };
 
 static const struct id_part faraday_part[] = {
