--- a/libtop.c	2022-11-04 08:11:37.000000000 +0800
+++ b/libtop.c	2023-01-02 17:23:13.355131191 +0800
@@ -710,6 +710,10 @@
 			size = SHARED_REGION_SIZE_ARM;
 		break;
 
+		case CPU_TYPE_ARM64:
+			base = SHARED_REGION_BASE_ARM64;
+			size = SHARED_REGION_SIZE_ARM64;
+		break;
 
 		case CPU_TYPE_X86_64:
 			base = SHARED_REGION_BASE_X86_64;
@@ -1344,6 +1348,7 @@
 	switch(type)
 	{
 		case CPU_TYPE_ARM:			return SHARED_REGION_SIZE_ARM;
+		case CPU_TYPE_ARM64:        return SHARED_REGION_SIZE_ARM64;
 		case CPU_TYPE_POWERPC:		return SHARED_REGION_SIZE_PPC;
 		case CPU_TYPE_POWERPC64:	return SHARED_REGION_SIZE_PPC64;
 		case CPU_TYPE_I386:			return SHARED_REGION_SIZE_I386;
diff -ur top-top-137/pid.c ./pid.c
--- top-top-137/pid.c	2022-11-04 08:11:37.000000000 +0800
+++ ./pid.c	2023-01-02 17:23:13.339775496 +0800
@@ -53,6 +53,11 @@
 	proc_is_foreign = true;
 #endif
 	break;
+    case CPU_TYPE_ARM64:
+#if !defined(__arm64__)
+    proc_is_foreign = true;
+#endif
+    // FALLTHROUGH
     case CPU_TYPE_ARM:
 #if !defined(__arm__)
 	proc_is_foreign = true;
