--- a/src/vamp-hostsdk/Files.cpp	1589799229.000000000
+++ b/src/vamp-hostsdk/Files.cpp	1709455092.067705833
@@ -302,20 +302,30 @@ Files::listFiles(string dir, string exte
 #endif
 #else
 
-    size_t extlen = extension.length();
-    DIR *d = opendir(dir.c_str());
-    if (!d) return files;
-            
-    struct dirent *e = 0;
-    while ((e = readdir(d))) {
- 
-        size_t len = strlen(e->d_name);
-        if (len < extlen + 2 ||
-            e->d_name + len - extlen - 1 != "." + extension) {
-            continue;
-        }
+    DIR *d = NULL;
+#if __APPLE__
+    // Torrekie: Try all possible exts
+    for (const auto& extensions : {"so", "dylib", "bundle"})
+#endif
+    {
+#if __APPLE__
+        extension = extensions;
+#endif
+        size_t extlen = extension.length();
+        d = opendir(dir.c_str());
+        if (!d) return files;
 
-        files.push_back(e->d_name);
+        struct dirent *e = 0;
+        while ((e = readdir(d))) {
+
+            size_t len = strlen(e->d_name);
+            if (len < extlen + 2 ||
+                e->d_name + len - extlen - 1 != "." + extension) {
+                continue;
+            }
+
+            files.push_back(e->d_name);
+        }
     }
 
     closedir(d);
--- a/src/vamp-hostsdk/PluginHostAdapter.cpp	1589799229.000000000
+++ b/src/vamp-hostsdk/PluginHostAdapter.cpp	1709453597.403621208
@@ -84,7 +84,7 @@ PluginHostAdapter::getPluginPath()
 #else
 #define PATH_SEPARATOR ':'
 #ifdef __APPLE__
-#define DEFAULT_VAMP_PATH "$HOME/Library/Audio/Plug-Ins/Vamp:/Library/Audio/Plug-Ins/Vamp"
+#define DEFAULT_VAMP_PATH "$HOME/Library/Audio/Plug-Ins/Vamp:/Library/Audio/Plug-Ins/Vamp:$HOME/vamp:$HOME/.vamp:/usr/local/lib/vamp:/usr/libexec/vamp:/usr/lib/vamp"
 #else
 #define DEFAULT_VAMP_PATH "$HOME/vamp:$HOME/.vamp:/usr/local/lib/vamp:/usr/lib/vamp"
 #endif
